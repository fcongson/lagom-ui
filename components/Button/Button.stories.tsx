import { action } from "@storybook/addon-actions";
import { Button } from "./Button";

export default {
  title: "Button",
  component: Button,
};

export const ButtonPrimary = () => (
  <Button onClick={action("clicked")} primary>
    Button primary
  </Button>
);

export const ButtonSecondary = () => (
  <Button onClick={action("clicked")}>Button secondary</Button>
);
