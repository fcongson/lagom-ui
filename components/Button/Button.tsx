import styled from "styled-components";
import {
  buttonStyle,
  ButtonStyleProps,
  typography,
  TypographyProps,
} from "styled-system";

const StyledButton = styled.button<ButtonStyleProps & TypographyProps>`
  ${typography}
  ${buttonStyle}

  display: flex;
  justify-content: center;
  align-items: center;

  height: 48px;
  width: fit-content;
  border: 1px solid black;
  border-radius: 24px;
  padding: 8px 24px;
  cursor: pointer;

  transition: transform 200ms ease-in-out;

  &:hover {
    transform: translateY(-2px);
  }

  &:active {
    transform: translateY(1px);
    transition: transform 100ms ease-in-out;
  }
`;

StyledButton.defaultProps = {
  fontFamily: "sansSerif",
  fontWeight: "link",
  fontSize: "body",
  lineHeight: 2,
};

interface ButtonProps extends React.ComponentPropsWithoutRef<"button"> {
  /**
   * Choose between the primary and secondary variants
   */
  primary?: boolean;
}

export const Button: React.FunctionComponent<ButtonProps> = ({
  children,
  primary = false,
  ...restProps
}) => (
  <StyledButton variant={primary ? "primary" : "secondary"} {...restProps}>
    {children}
  </StyledButton>
);
