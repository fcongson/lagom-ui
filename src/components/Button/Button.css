.lagom-button {
  display: flex;
  justify-content: center;
  align-items: center;

  height: 48px;
  width: fit-content;
  border: 1px solid;
  padding: 8px 24px;
  cursor: pointer;

  transition: transform 200ms ease-in-out;

  border-radius: var(--lagom-component-button-border-radius);
  font-family: var(--lagom-core-font-families-sans-serif);
  font-weight: var(--lagom-core-font-weights-semi-bold);
  font-size: var(--lagom-core-font-sizes-body);
}

.lagom-button--primary {
  background-color: var(--lagom-component-button-color-primary-background);
  color: var(--lagom-component-button-color-primary-text);
  border-color: var(--lagom-component-button-color-primary-border);
}

.lagom-button--secondary {
  background-color: var(--lagom-component-button-color-secondary-background);
  color: var(--lagom-component-button-color-secondary-text);
  border-color: var(--lagom-component-button-color-secondary-border);
}

.lagom-button:hover {
  transform: translateY(-2px);
}

.lagom-button:active {
  transform: translateY(1px);
  transition: transform 100ms ease-in-out;
}
