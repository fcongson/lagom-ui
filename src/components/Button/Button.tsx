import styled, { useTheme } from "styled-components";
import {
  border,
  BorderProps,
  buttonStyle,
  ButtonStyleProps,
  typography,
  TypographyProps,
} from "styled-system";
import { theme } from "../../themes/theme";

const StyledButton = styled.button<
  ButtonStyleProps & TypographyProps & BorderProps
>`
  ${typography}
  ${border}
  ${buttonStyle}

  display: flex;
  justify-content: center;
  align-items: center;

  height: 48px;
  width: fit-content;
  border: 1px solid;
  /* border-radius: 24px; */
  padding: 8px 24px;
  cursor: pointer;

  transition: transform 200ms ease-in-out;

  &:hover {
    transform: translateY(-2px);
  }

  &:active {
    transform: translateY(1px);
    transition: transform 100ms ease-in-out;
  }
`;

StyledButton.defaultProps = {
  theme,
  fontFamily: "sansSerif",
  fontWeight: "link",
  fontSize: "body",
  lineHeight: 2,
  borderColor: "black",
  borderRadius: 24,
};

interface ButtonProps extends React.ComponentPropsWithoutRef<"button"> {
  /**
   * Choose between the primary and secondary variants
   */
  primary?: boolean;
}

export const Button: React.FunctionComponent<ButtonProps> = ({
  children,
  primary = false,
  ...restProps
}) => {
  const providedTheme = useTheme();
  const mergedTheme = { ...theme, ...providedTheme };
  return (
    <StyledButton
      theme={mergedTheme}
      variant={primary ? "primary" : "secondary"}
      {...restProps}
    >
      {children}
    </StyledButton>
  );
};
