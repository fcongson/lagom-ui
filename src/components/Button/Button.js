import { clsx } from "clsx";
import React from "react";

import "./Button.css";

export const Button = ({
  children,
  className,
  primary = false,
  ...restProps
}) => (
  <button
    className={clsx(
      "lagom-button",
      className,
      primary ? "lagom-button--primary" : "lagom-button--secondary"
    )}
    {...restProps}
  >
    {children}
  </button>
);
