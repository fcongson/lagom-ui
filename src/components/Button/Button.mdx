import {
  Meta,
  Primary,
  Controls,
  Story,
  Canvas,
  Unstyled,
} from "@storybook/blocks";
import * as ButtonStories from "./Button.stories";

<Meta of={ButtonStories} />

<Unstyled>

# Button

A button is ...

<Primary />

## Props

<Controls />

## Stories

### Primary

A button can be of primary importance.

<Canvas of={ButtonStories.Primary} />

### Secondary

A button can be of secondary importance.

<Canvas of={ButtonStories.Secondary} />

### Custom

A button can be a custom variant.

<Canvas of={ButtonStories.Custom} />

## Experimental

### With Icon

A button can include an icon.

<Canvas of={ButtonStories.WithExperimentalIcon} />

</Unstyled>
