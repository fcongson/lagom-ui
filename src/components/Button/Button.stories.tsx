import { action } from "@storybook/addon-actions";
import { Meta, StoryObj } from "@storybook/react";
import { Button } from "./Button";

const meta: Meta = {
  title: "Components/Button",
  component: Button,
  // tags: ["autodocs"],
  args: {
    onClick: action("clicked"),
    primary: true,
    children: "Button primary",
  },
};

export default meta;

type Story = StoryObj<typeof Button>;

export const Primary: Story = {};

export const Secondary: Story = {
  ...Primary,
  args: {
    primary: false,
    children: "Button secondary",
  },
};

export const Custom: Story = {
  ...Primary,
  args: {
    primary: false,
    style: {
      backgroundColor: "var(--lagom-core-colors-olive-base)",
      borderColor: "var(--lagom-core-colors-olive-base)",
      color: "var(--lagom-core-colors-white)",
      borderRadius: 4,
      fontFamily: "monospace",
    },
    children: "Button custom",
  },
};

export const WithExperimentalIcon: Story = {
  ...Primary,
  args: {
    primary: false,
    children: "With icon",
    icon: "ðŸš€",
  },
};
