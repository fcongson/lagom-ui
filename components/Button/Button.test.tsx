import { shallow } from "enzyme";
import { Button } from "./Button";

describe("Button", () => {
  it("Renders", () => {
    const button = shallow(<Button>Text</Button>);
    expect(button).toMatchSnapshot();
  });
});
