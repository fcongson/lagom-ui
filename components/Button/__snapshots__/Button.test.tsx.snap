// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Renders 1`] = `
<button
  className="sc-bdnxRM eBJmcI"
>
  Text
</button>
`;
